File: Leetcode/graph/1202_SmallestStringWithSwaps/Solution.java
class Solution {
    public String smallestStringWithSwaps(String s, List<List<Integer>> pairs) {
        int n = s.length();
        boolean[] visited = new boolean[n];
        List<List<Integer>> graph = new ArrayList<>();
        for(int i = 0; i<n; ++i) graph.add(new ArrayList<Integer>());
        for(int i = 0; i<pairs.size(); ++i){
            graph.get(pairs.get(i).get(1)).add(pairs.get(i).get(0));
            graph.get(pairs.get(i).get(0)).add(pairs.get(i).get(1));
        }
        List<Character> res = s.chars()                 // IntStream
                                .mapToObj(e -> (char)e)     // Stream<Character>
                                .collect(Collectors.toList());
        for(int i = 0; i<n; ++i){
            if(visited[i])
                continue;
            List<Integer> component = new ArrayList<Integer>();
            dfs(i, component, visited, graph);
            Collections.sort(component);
            List<Character> chars = new ArrayList<>();
            for(int num: component)
                chars.add(res.get(num));
            Collections.sort(chars);
            for(int j = 0; j<component.size(); ++j)
                res.set(component.get(j), chars.get(j));
        }
        StringBuilder b = new StringBuilder();
        res.forEach(b::append);
        return b.toString();
    }
    private void dfs(int src, List<Integer> component,
                     boolean[] visited, List<List<Integer>> graph){
        visited[src] = true;
        component.add(src);
        for(int dest: graph.get(src)){
            if(!visited[dest])
                dfs(dest, component, visited, graph);
        }
    }
}

File: Leetcode/graph/1202_SmallestStringWithSwaps/Solution.py
class Solution(object):
    def smallestStringWithSwaps(self, s, pairs):
        """
        :type s: str
        :type pairs: List[List[int]]
        :rtype: str
        """
        def dfs(i):
            visited[i] = True
            component.append(i)
            for j in adj_lst[i]:
                if not visited[j]:
                    dfs(j)

        n = len(s)
        adj_lst = [[] for _ in range(n)]
        for i, j in pairs:
            adj_lst[i].append(j)
            adj_lst[j].append(i)
        visited = [False for _ in range(n)]
        res = list(s)
        for i in range(n):
            if not visited[i]:
                component = []
                dfs(i)
                #print(component)
                component.sort()
                #print(component)
                chars = [res[k] for k in component]
                #print(chars)
                chars.sort()
                #print(chars)
                for i in range(len(component)):
                    res[component[i]] = chars[i]
        return ''.join(res)
